refactor: rpc send transfer

fbwoolf opened the pull request:
> Try out Leather build d9a6685 — [Extension build](https://github.com/leather-io/extension/actions/runs/13661521684), [Test report](https://leather-io.github.io/playwright-reports/refactor/rpc-send-transfer), [Storybook](https://refactor/rpc-send-transfer--65982789c7e2278518f189e7.chromatic.com), [Chromatic](https://www.chromatic.com/library?appId=65982789c7e2278518f189e7&branch=refactor/rpc-send-transfer)<!-- Sticky Header Marker -->

This PR did a deep dive into the Approver UX implementation for `sendTransfer`, all of the flow was tightly integrated into one`pages` flow w/out the ability to really share any of it with other rpc flows. This first PR mostly pulls out all of the work on the new `Edit fees` step and allows it be shared with other flows and b/w Stacks and Bitcoin. I may need to do more here with all the `ApproveTransaction...` components, but I wanted to pull the fees code out first. I did make a new `RpcSendTransferContext` that extends `RpcRequestContext`. There might be more to share here when I start to develop the `stx_callContract` flow.

I created a separate context to manage fees so that it can be used with the send flows too, etc. I named the new feature `FeeEditor` but open to feedback on it. The context wraps the rpc context, but open to feedback on that too.

The container now looks like...
```js
<BitcoinUtxosLoader>
  {utxos => (
    <BitcoinFeeEditorProvider
      amount={amount}
      availableBalance={btcBalance.balance.availableBalance}
      isSendingMax={false}
      marketData={btcMarketData}
      recipients={recipients}
      utxos={utxos}
    >
      <RpcSendTransferProvider
        value={{
          ...rpcSendTransferContext,
          isLoading: btcBalance.isLoadingAllData,
          onUserActivatesFeeEditor: () => navigate(RouteUrls.EditFee),
          onUserActivatesSwitchAccount: toggleSwitchAccount,
          utxos,
        }}
      >
        <Outlet />
      </RpcSendTransferProvider>
    </BitcoinFeeEditorProvider>
  )}
</BitcoinUtxosLoader>
```

Some things I noticed during the refactor:
- Files were located in the wrong place, new fees files for rpc were being located with old send form fees
- Functions were being copied and duplicated instead of shared
- Multiple props were being used to handle the same thing and/or trying to fix side effects
- There was no thought put into making the fees step shareable b/w flows and/or chains

I also uncovered a bug with the custom input which was flickering in the UI and allowed an undefined input...
<img width="406" alt="Screenshot 2025-03-03 at 12 23 50 PM" src="https://github.com/user-attachments/assets/94b40116-32f1-48aa-8350-1e861bae967f" />
<img width="411" alt="Screenshot 2025-03-03 at 12 24 05 PM" src="https://github.com/user-attachments/assets/beca07f9-3ac8-4de9-953c-efe19d2fb72b" />

Fixed... (which maybe we should validate and prevent?)
![Screenshot 2025-03-03 at 12 25 20 PM](https://github.com/user-attachments/assets/ea34303d-ea74-40e4-9b82-de1ed7a73060)
![Screenshot 2025-03-03 at 12 24 22 PM](https://github.com/user-attachments/assets/d7a0940e-86f7-4f9c-a9d7-046588bf06a9)

FeeEditor trigger component...
![Screenshot 2025-03-03 at 12 23 24 PM](https://github.com/user-attachments/assets/125af6cd-5ffb-4619-a988-1e9d99fe9a6c)

FeeEditor...
![Screenshot 2025-03-03 at 12 22 42 PM](https://github.com/user-attachments/assets/743d5630-b642-45de-b445-f19925ca7971) 
![Screenshot 2025-03-03 at 12 23 05 PM](https://github.com/user-attachments/assets/d10b5e2d-c363-42ba-8d41-2edc4d5f5084)

fbwoolf commented:
Test transaction was successful here selecting the fast fee: https://mempool.space/tx/a3ab3846c74095ce18f9d1c15f896cd63fe86ed791eb34499781f4e655dade08

fbwoolf commented on src/app/features/fee-editor/use-fee-editor.ts:
```diff
@@ -0,0 +1,43 @@
+  getCustomEditorFee(rate: number): EditorFee;
+}
+export function useFeeEditor({ editorFees, getCustomEditorFee }: UseFeeEditorArgs) {
+  const [currentEditorFee, setCurrentEditorFee] = useState<EditorFee | null>(
```
Looking at this setup now bc came from `useFees` (see deleted file). The tests are failing bc the fees stay in loading state. I had a console.log that was helping to trigger a rerender or this remained `null`. I think I'll have to use a loader component or something or we have to use `useEffects` to set state which like below, which isn't good.

kyranjamie replied:
+1 to loaders or another kind of guard that prevents need for useEffect

fbwoolf replied:
My last commit refactors to use loaders and removes the two hooks causing issues.

kyranjamie commented on src/app/features/fee-editor/fee-editor.context.ts:
```diff
@@ -0,0 +1,39 @@
+
+import type { MarketData, Money } from '@leather.io/models';
+
+export type EditorFeeType = 'slow' | 'standard' | 'fast' | 'custom';
```
FeePriority?

kyranjamie commented on src/app/features/fee-editor/fee-editor.context.ts:
```diff
@@ -0,0 +1,39 @@
+
+export type EditorFeeType = 'slow' | 'standard' | 'fast' | 'custom';
+
+export interface EditorFee {
+  type: EditorFeeType;
+  feeRate: number | null;
+  feeValue: Money | null;
+  time: string;
+}
```
`FeeEditor` is a great name for this component. Calling fee data within it `EditorFee` I find a bit hard to follow.

`FeeSelection`, `SelectedFee`, `FeeEdtiorSelection`?

kyranjamie commented on src/app/features/fee-editor/fee-editor.tsx:
```diff
@@ -0,0 +1,77 @@
+  function onCancel() {
+    onSetCustomEditorFeeRate(selectedEditorFee?.feeRate?.toString() || '');
+    onSetSelectedEditorFee(currentEditorFee);
+    navigate(-1);
```
I can never trust a nav -1. Is this tied to routes at all? If so would `../` make more sense? Or just have the action passed in?

kyranjamie commented on src/app/features/fee-editor/fee-editor.tsx:
```diff
@@ -0,0 +1,77 @@
+    editorFees,
+    selectedEditorFee,
+    getCustomEditorFee,
+    onSetCurrentEditorFee,
+    onSetCustomEditorFeeRate,
```
I wonder if it'd be cleaner to have a single fee setting, using a discriminating string literal?

e.g. 
```ts
// making up prop names here
onSetCurrentEditorFee({ type: 'custom', customValue: 'xxx' });
onSetCurrentEditorFee({ type: 'high' });
```

kyranjamie commented on src/app/features/fee-editor/fee-editor.utils.ts:
```diff
@@ -0,0 +1,32 @@
+    captionLeft: time,
+    titleRight: feeValue ? formatMoneyPadded(feeValue) : 'N/A',
+    captionRight: feeValue
+      ? `${feeRate} ${feeValue.symbol === 'BTC' ? 'sats' : 'uSTX'}/vB · ${i18nFormatCurrency(
```
Rather than ternaries, could we have a `Record<CryptoSymbol, Unit>` we ref?

kyranjamie reviewed (approved):
Fantastic refactor @fbwoolf. We need to use this in many places, so it needs to be uncoupled from any specific RPC endpoint

kyranjamie commented on src/app/components/loaders/bitcoin-utxos-loader.tsx:
```diff
@@ -1,11 +1,14 @@
-  const { data: utxos = [] } = useCurrentNativeSegwitUtxos();
+  const { data: utxos = [], filteredUtxosQuery } = useCurrentNativeSegwitUtxos();
+  // Forcing a refetch to ensure UTXOs are fresh
+  useOnMount(() => filteredUtxosQuery.refetch());
```
Wouldn't using `refreshOnMount` do the same here and not require the manual refresh?

fbwoolf replied:
@alter-eggo I just relocated this from I believe something you had originally in `useRpcSendTransfer`, do you know if this is even really needed?

kyranjamie commented on src/app/features/fee-editor/fee-editor.context.ts:
```diff
@@ -11,29 +11,27 @@ export interface EditorFee {
   onSetCustomEditorFeeRate(value: string | null): void;
-  onSetCurrentEditorFee(value: EditorFee | null): void;
-  onSetSelectedEditorFee(value: EditorFee | null): void;
+  onSetCurrentEditorFee(value: EditorFee): void;
+  onSetSelectedEditorFee(value: EditorFee): void;
```
From name alone, unclear what the difference between these two is

alter-eggo replied:
user might select fee type, but not save the selected value and just go back
may be better `onSaveSelectedEditorFee` ?

fbwoolf replied:
Yeah, the editor view has it's own need to track what the user is selecting, and then once they have 'saved' it that becomes current. We need to track it just in case they cancel. We have a similar need with editing a nonce but we call it `setLoadedNonce`. I'll work on renaming some things with my next pass through the code today.

fbwoolf replied:
I was able to simplify the state/naming in my last commit. There was just so much going on it was hard to track. I ended up just keeping track of a loaded fee in case the user cancels, but then it seems fine just to keep track of the `selectedFee`.

kyranjamie commented on src/app/pages/rpc-send-transfer/use-rpc-send-transfer-actions.tsx:
```diff
@@ -18,8 +18,14 @@ import { useRpcSendTransferContext } from './rpc-send-transfer.context';
-  const { amountAsMoney, isLoading, recipients, requestId, tabId, utxos } =
-    useRpcSendTransferContext();
+  const {
+    amount: amountAsMoney,
```
Can this now just be `amount`?

fbwoolf replied:
Yep, I'll change, was going that route but wasn't sure if people are ok with `amount.amount` bc of the Money type.

kyranjamie commented on src/app/features/fee-editor/components/custom-fee-item.tsx:
```diff
@@ -5,33 +5,37 @@ import { Stack } from 'leather-styles/jsx';
+      onClick={() => onSetSelectedEditorFee(fee)}
       variant="outline"
       borderWidth={isSelected ? '2px' : '1px'}
       borderColor={isSelected ? 'ink.border-selected' : 'ink.border-default'}
-      // Add margin compensation when not selected to maintain consistent size
+      // Add margin compensation to maintain consistent size
```
Ideally, this would've been implemented with pseudo elements the same way `Input` was to avoid any box content overflows when applying borders.

fbwoolf replied:
Yep, I can try to look at this too today.

fbwoolf replied:
I'll create an issue for this since it is a change in the UI library.

fbwoolf replied:
https://linear.app/leather-io/issue/LEA-2234/refactor-ui-lib-button-for-web-to-handle-border-on-selected

alter-eggo reviewed (approved):
this is great work @fbwoolf 
tested and fee selection works ok for me, tx was also submitted with proper fee

only thing - shouldn't we set default custom fee value to a standart one?

![Screenshot 2025-03-04 at 14 53 08](https://github.com/user-attachments/assets/47b7b0af-d24c-49c4-8e96-8ac839e13f59)

fbwoolf commented:
> 
> only thing - shouldn't we set default custom fee value to a standart one?
> 
Thx for flagging @alter-eggo, I thought that was working so maybe I broke with my last commit. I'll fix today.

fbwoolf commented:
I think I have addressed all code comments, and had a successful tx here: https://mempool.space/tx/b2ee0c28f38140e9be46754731427dda641231358d4cb98f5c2722454088c386

Going to merge so I can move fwd with call contract.
